exp/mutators/azureasomanagedcontrolplane.go

```diff
@@ -165,3 +177,172 @@ func setManagedClusterPodCIDR(ctx context.Context, cluster *clusterv1.Cluster, m
+	if client.IgnoreNotFound(err) != nil {
+		return err
+	}
+	if len(getMC.Status.AgentPoolProfiles) != 0 {
+		return nil
+	}
+
+	log.V(4).Info("gathering agent pool profiles to include in ManagedCluster create")
+	agentPools, err := agentPoolsFromManagedMachinePools(ctx, ctrlClient, cluster.Name, namespace)
+	if err != nil {
+		return err
+	}
+	mc, err := ctrlClient.Scheme().New(managedCluster.GroupVersionKind())
+	if err != nil {
+		return err
+	}
+	err = ctrlClient.Scheme().Convert(managedCluster, mc, nil)
+	if err != nil {
+		return err
+	}
+	setAgentPoolProfiles.val = fmt.Sprintf("<slice of length %d>", len(agentPools))
+	logMutation(log, setAgentPoolProfiles)
+	err = setAgentPoolProfilesFromAgentPools(mc.(conversion.Convertible), agentPools)
+	if err != nil {
+		return err
+	}
+	err = ctrlClient.Scheme().Convert(mc, managedCluster, nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func agentPoolsFromManagedMachinePools(ctx context.Context, ctrlClient client.Client, clusterName string, namespace string) ([]conversion.Convertible, error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "mutators.agentPoolsFromManagedMachinePools")
+	defer done()
+
+	asoManagedMachinePools := &infrav1exp.AzureASOManagedMachinePoolList{}
+	err := ctrlClient.List(ctx, asoManagedMachinePools,
+		client.InNamespace(namespace),
+		client.MatchingLabels{
+			clusterv1.ClusterNameLabel: clusterName,
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list AzureASOManagedMachinePools: %w", err)
+	}
+
+	var agentPools []conversion.Convertible
+	for _, asoManagedMachinePool := range asoManagedMachinePools.Items {
+		resources, err := ApplyMutators(ctx, asoManagedMachinePool.Spec.Resources)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, u := range resources {
+			if u.GroupVersionKind().Group != asocontainerservicev1hub.GroupVersion.Group ||
+				u.GroupVersionKind().Kind != "ManagedClustersAgentPool" {
+				continue
+			}
+
+			agentPool, err := ctrlClient.Scheme().New(u.GroupVersionKind())
+			if err != nil {
+				return nil, fmt.Errorf("error creating new %v: %w", u.GroupVersionKind(), err)
+			}
+			err = ctrlClient.Scheme().Convert(u, agentPool, nil)
+			if err != nil {
+				return nil, err
+			}
+
+			agentPools = append(agentPools, agentPool.(conversion.Convertible))
+			break
+		}
+	}
+
+	return agentPools, nil
+}
+
+func setAgentPoolProfilesFromAgentPools(managedCluster conversion.Convertible, agentPools []conversion.Convertible) error {
+	hubMC := &asocontainerservicev1hub.ManagedCluster{}
+	err := managedCluster.ConvertTo(hubMC)
+	if err != nil {
+		return err
+	}
+	hubMC.Spec.AgentPoolProfiles = nil
+
+	for _, agentPool := range agentPools {
+		hubPool := &asocontainerservicev1hub.ManagedClustersAgentPool{}
+		err := agentPool.ConvertTo(hubPool)
+		if err != nil {
+			return err
+		}
+
+		profile := asocontainerservicev1hub.ManagedClusterAgentPoolProfile{
```

Comment:
How much ongoing maintenance will this manual conversion take?

Reply:
All we should need to do is add any new AKS API versions for each new ASO release to the scheme we create in `main`. If the hub type happens to change between ASO releases, we should get a compiler error here. When that happens, we would need to update these fields to include any new fields in the new hub version.

If https://github.com/Azure/azure-service-operator/issues/2791 gets resolved, then none of that should be necessary.

Reply:
compiler error is the news I came to hear